src/client/QXmppOmemoManager.cpp

```diff
@@ -0,0 +1,5402 @@
+        error.type = QXmpp::SendError::EncryptionError;
+        reportFinishedResult(interface, { error });
+    } else {
+        std::optional<TrustLevels> acceptedTrustLevels;
+
+        if (params) {
+            acceptedTrustLevels = params->acceptedTrustLevels();
+        }
+
+        if (!acceptedTrustLevels) {
+            acceptedTrustLevels = ACCEPTED_TRUST_LEVELS;
+        }
+
+        auto future = d->encryptStanza(iq, { QXmppUtils::jidToBareJid(iq.to()) }, *acceptedTrustLevels);
+        await(future, this, [=, iq = std::move(iq)](std::optional<QXmppOmemoElement> omemoElement) mutable {
+            if (!omemoElement) {
+                QXmpp::SendError error;
+                error.text = QStringLiteral("OMEMO element could not be created");
+                error.type = QXmpp::SendError::EncryptionError;
+                reportFinishedResult(interface, { error });
+            } else {
+                QXmppOmemoIq omemoIq;
+                omemoIq.setId(iq.id());
+                omemoIq.setType(iq.type());
+                omemoIq.setLang(iq.lang());
+                omemoIq.setFrom(iq.from());
+                omemoIq.setTo(iq.to());
+                omemoIq.setOmemoElement(*omemoElement);
+
+                QByteArray serializedEncryptedIq;
+                QXmlStreamWriter writer(&serializedEncryptedIq);
+                omemoIq.toXml(&writer);
+
+                reportFinishedResult(interface, { serializedEncryptedIq });
+            }
+        });
+    }
+
+    return interface.future();
+}
+
+QFuture<QXmppE2eeExtension::IqDecryptResult> QXmppOmemoManager::decryptIq(const QDomElement &element)
+{
+    QFutureInterface<QXmppE2eeExtension::IqDecryptResult> interface(QFutureInterfaceBase::Started);
+
+    if (!d->isStarted) {
+        QXmpp::SendError error;
+        error.text = QStringLiteral("OMEMO manager must be started before decrypting");
+        error.type = QXmpp::SendError::EncryptionError;
+        reportFinishedResult(interface, { error });
+    } else if (QXmppOmemoIq::isOmemoIq(element)) {
+        QXmppOmemoIq omemoIq;
+        omemoIq.parse(element);
+
+        auto future = d->handleReceivedStanza(std::move(omemoIq));
+        await(future, this, [=](std::optional<QXmppOmemoIq> optionalDecryptedIq) mutable {
+            if (optionalDecryptedIq) {
+                QByteArray serializedDecryptedIq;
+                QXmlStreamWriter writer(&serializedDecryptedIq);
+
+                // The OMEMO element is removed by creating a QXmppIq out of the
+                // QXmppOmemoIq.
+                // Thus, only its remaining elements / extensions are
+                // serialized.
+                QXmppIq(*optionalDecryptedIq).toXml(&writer);
+                QDomDocument document;
+                document.setContent(serializedDecryptedIq, true);
+
+                reportFinishedResult(interface, { document.documentElement() });
```

Comment:
This is a bit ugly, you decrypt the omemo iq giving you XML as QByteArray and then (inside handleReceivedStanza()) you parse that into an OmemoIq, of which you then discard the Omemo part (copy into QXmppIq()) and serialize it to XML bytes again.

1. QXmppOmemoIq is parsed
1. OmemoElement is decrypted => QByteArray
1. QByteArray is parsed into QDomElement
1. QDomElement is parsed into QXmppOmemoIq (and by that the decrypted extension is parsed into QXmppElement)
1. QXmppOmemoIq is copied into QXmppIq
1. QXmppIq (including QXmppElement extension) is serialized into QByteArray
1. QByteArray is parsed into QDomElement
1. QDomElement is further processed (and then parsed into the right IQ type)

Especially the unnecessary parsing + serialization of the IQ isn't nice. This comes from an architecture that was only designed for messages not for IQs. Please reduce it to this:

1. QXmppOmemoIq is parsed
1. OmemoElement is decrypted => QByteArray
1. QByteArray is parsed into QDomElement (here either the QByteArray or the QDomElement needs to be adapted to have an IQ tag with from/to/type)
1. QDomElement is further processed (and then parsed into the right IQ type)